### ISSUE TITLE
Missing TEST_ASSERT_NOT_EQUAL_{PTR,STRING,MEMORY} macros

### ISSUE THREAD

**Issue by nopeslide** [ThrowTheSwitch/Unity#491]
A lot of dynamic structures use self-referencing when empty.
A macro like `TEST_ASSERT_NOT_EQUAL_PTR` enables simple empty/not-empty checks for such structures.

see linuxs list implementation as an example:
```
/**
 * list_empty - tests whether a list is empty
 * @head: the list to test.
 */
static inline int list_empty(const struct list_head *head)
{
	return READ_ONCE(head->next) == head;
}
```

Further negated macros for strings and memory would enable checking if some memory areas were touched.

**Comment by laurencedv** [comment]
I agree with the last point, being able to assert if some chunk of memory is not at the init value, but whichever it's value would be very useful. :)

**Comment by AlariOis** [comment]
How should these macros act in case of NULL arguments?
FAIL if actual = NULL and expected = NULL (because they are equal)
SUCCEED if actual = NULL and expected = NOT NULL 
SUCCEED if actual = NOT NULL and expected = NULL 

Or would it make more sense to still FAIL in case actual is NULL and expected is not because there is no actual data to compare?

**Comment by nopeslide** [comment]
All of them should behave like the non-inverted versions.
I don't know if the non-inverted versions do any argument checks.

**Comment by mvandervoord** [comment]
The non-inverted variants pass if both are NULL, therefore I would expect this to fail if both are NULL and pass if either is NULL.

**Comment by nopeslide** [comment]
> The non-inverted variants pass if both are NULL, therefore I would expect this to fail if both are NULL and pass if either is NULL.

sounds like

**Comment by nopeslide** [comment]
sorry, I hit the wrong button.

**Comment by alecoding** [comment]
I would expect these macros to do just what they are supposed to do by reading the name:

TEST_ASSERT_NOT_EQUAL_PTR: pass if the pointers differ, no matter if one is NULL or not.  If they are both NULL the macros should report a failure

TEST_ASSERT_NOT_EQUAL_STRING: compare the strings no matter what the pointer is.  If it is NULL too bad.  The tester should add a pointer check before if concerned that it could be not good

TEST_ASSERT_NOT_EQUAL_MEMORY: same as TEST_ASSERT_NOT_EQUAL_STRING

**Comment by mvandervoord** [comment]
In the latter two cases, @alecoding I disagree. Where possible, Unity tries to identify issues that the developer has run into. A common problem is passing a null pointer where they intended to pass a block of memory or a string. By forcing them to remember to add a TEST_ASSERT_NOT_NULL before every string or memory assertion, we're allowing the tests to mysteriously crash instead of pointing out an easily identifiable and common problem.

**Comment by AlariOis** [comment]
@mvandervoord, would be awesome if you would check out my PR about this.
I added this on top of another feature in hopes of saving future work but this was maybe not the best idea on my part.
I could make a new PR that addresses only this issue specifically if needed.

**Comment by alecoding** [comment]
> In the latter two cases, @alecoding I disagree. Where possible, Unity tries to identify issues that the developer has run into. A common problem is passing a null pointer where they intended to pass a block of memory or a string. By forcing them to remember to add a TEST_ASSERT_NOT_NULL before every string or memory assertion, we're allowing the tests to mysteriously crash instead of pointing out an easily identifiable and common problem.

I believe this is a personal preference.  I'm not sure that checking for NULL is always necessary and maybe correct. There might be some extreme cases in which NULL is actually a valid pointer.  Again, as a personal preference I prefer a set of minimal checks rather than macros that performs multiple operations.  My view is that in the long term a  separate macros for separate problems provide more transparency and flexibility

**Comment by rstahn** [comment]
>  I'm not sure that checking for NULL is always necessary and maybe correct. There might be some extreme cases in which NULL is actually a valid pointer.
That is a point that came to my mind as well. Especially in Embedded Environments it is not unusual that RAM or FLASH starts at address 0. Hidden NULL-pointer checks could then lead to unexpected results, when NULL is actually a legal pointer.

Furthermore I wonder, why we care so much about NULL pointers. Even when I consider it a "forbidden" value, it is only a single corner case of an invalid pointer. When I think about invalid pointers e.g. created by memory corruption, they usually are not equal to NULL. I guess the "obsession" about NULL is only valid, if some kind of dynamic memory management (like malloc()) is used where a NULL pointer presents an error value in case a memory block could not be allocated. In embedded environments dynamic memory management is typically avoided, though.

Therefore, I would prefer any "hidden" NULL pointer checks to be optional.
